User can check out a book

margaretoti opened the pull request:
A user can checkout (also known as borrow) a book from the library. Checked out books become unavailable for others to checkout. 

<img width="469" alt="screen shot 2017-04-11 at 10 09 15 am" src="https://cloud.githubusercontent.com/assets/9501674/24913386/0f9418c0-1e9f-11e7-9707-79391ad07f2a.png">

Note on screen shot: Unavailable books are those that have been checked out.

margaretoti committed:
User can checkout a book

derekprior commented on spec/models/book_spec.rb:
```diff
@@ -6,6 +6,11 @@
+  end
+
+  describe 'associations' do
+    it { should have_many(:checkouts) }
```
I usually wouldn't bother with specs that assert that an association exists. If the association is useful, it'll likely be covered by some feature spec, as it is here. That said, some people like these and if that's the case I don't feel strongly that you should remove them. Just thought I'd mention it.

margaretoti replied:
Thanks for the advice @derekprior. If you don't mind, I am going to keep them for now. Once I get a better hang of TDD, I will start to remove them.

derekprior commented on app/controllers/checkouts_controller.rb:
```diff
@@ -0,0 +1,10 @@
+class CheckoutsController < ApplicationController
+  before_action :require_login, :allow_anyone
```
What does `allow_anyone` do and how is it different from not having that declared at all?

margaretoti replied:
@derekprior, `allow_anyone` and `require_admin` serve to check (and test) who is allowed to access what book routes. Users (patrons of the library) can only view books and check them out. Admin have the ability to add, delete, and edit books (`admin/books controller`), and can also enjoy the regular benefits of the library - viewing books and checking them out. Thus `books controller` has `allow_anyone`, while `admin/books controller` has `require_admin`.

derekprior replied:
Ok, I see now... You have an after_filter that makes sure authorization has been checked and allow_anyone records that it has.

margaretoti replied:
Exactly!

derekprior commented on app/models/book.rb:
```diff
@@ -1,4 +1,15 @@
+  end
+
+  def self.unavailable
+    includes(:checkouts).where.not(checkouts: { id: nil })
```
Nice. These ended up being much simpler in ActiveRecord than I had thought they would be. Good job.

margaretoti replied:
@derekprior, It actually turned out to be quite simple. Just so you know, I think it is only supported in Rails 4 and up.

derekprior commented on db/migrate/20170410185513_create_checkouts.rb:
```diff
@@ -0,0 +1,9 @@
+    create_table :checkouts do |t|
+      t.timestamps null: false
+      t.references :book, index: true, foreign_key: true, null: false
+      t.references :user, index: true, foreign_key: true, null: false
```
How do we ensure that a book can't be checked out twice? 

The UI tries to prevent this by not showing a checkout link for unavailable books, but if two users load that page at the same time, they'll both see a book as available and be able to check it out. We should be able to prevent this at the database and model level with validations and constraints/indexes.

margaretoti replied:
@derekprior, This is a very good point. I need to think about it, but I should have a solution shortly.

derekprior commented on spec/features/user_checks_out_book_spec.rb:
```diff
@@ -0,0 +1,25 @@
+    visit books_path(as: patron)
+
+    expect(page).to have_available_book(book.title)
+    expect(page).not_to have_checked_out_book(book.title)
```
These read really nicely! Good job.

margaretoti committed:
Add uniqueness constraint to book id

A uniqueness constraint was added to the book id column of the checkouts
table to ensure that a different user can not checkout a book that has
been checked out. Before this change, the UI alone was in charge of
preventing the same book from being checked out by two different users.

derekprior commented on db/migrate/20170410185513_create_checkouts.rb:
```diff
@@ -0,0 +1,12 @@
+    end
+
+    remove_index :checkouts, :book_id
+    add_index :checkouts, :book_id, unique: true
```
You can avoid the `remove_index`, `add_index` combination here by updating line 5 like so:

```ruby
t.references :book, index: { unique: true }, ...
```

derekprior reviewed (approved):
Looks good. just one small comment on the migration
